File: lib/discount.rb
class Discount < Errol::Entity
  class Null
    def value
      Money.new(0)
    end

    def nil?
      true
    end

    def expired?
      false
    end

    def pending?
      false
    end

    def code
      ''
    end
  end
  def self.null
    Null.new
  end
  require_relative './discount/record'
  entry_accessor  :code,
                  :value,
                  :allocation,
                  :customer_allocation,
                  :start_datetime,
                  :end_datetime

  def expired?
    end_datetime < DateTime.now
  end

  def pending?
    start_datetime > DateTime.now
  end

  def number_of_redemptions
    ShoppingBaskets.count(:checked_out => true, :discount => self)
  end
end

File: apps/sales/lib/sales/discount.ex
defmodule UM.Sales.Discount do
  defmodule Zero do
    defstruct [id: nil]
  end
  defstruct [
    id: nil,
    code: nil,
    value: nil,
    allocation: nil,
    customer_allocation: nil,
    start_datetime: nil,
    end_datetime: nil
  ]

  def zero, do: %Zero{}

  def value(%Zero{}), do: 0
  def value(%{value: value}), do: value

  def check_availabile(%Zero{}, _) do
    {:ok, %Zero{}}
  end

  def code() do
    code
  end
  # pending?
  # expired?
end
